### ISSUE TITLE
Composed modules not working on YARN

### ISSUE THREAD

**Issue by clemper** [spring-projects/spring-xd#1704]
I am trying to use composed modules when running on YARN.  

In ZK, each child module definition of the composed module gets serialized as follows:

```
{"@class":"org.springframework.xd.module.SimpleModuleDefinition","name":"transform","type":"processor","location":"file:/tmp/hadoop-hduser/nm-local-dir/usercache/hduser/appcache/application_1433789137218_0001/filecache/17/spring-xd-yarn-1.1.2.RELEASE.zip/modules/processor/transform/"}
```

When I try to use the composed module on YARN, it may be deployed to a different container where the "location" file path is not valid.  In this case I get the following exception:

```
java.lang.IllegalArgumentException: File must exist
	at org.springframework.boot.loader.data.RandomAccessDataFile.<init>(RandomAccessDataFile.java:67)
	at org.springframework.boot.loader.data.RandomAccessDataFile.<init>(RandomAccessDataFile.java:51)
	at org.springframework.boot.loader.jar.JarFile.<init>(JarFile.java:95)
	at org.springframework.boot.loader.archive.JarFileArchive.<init>(JarFileArchive.java:61)
	at org.springframework.boot.loader.archive.JarFileArchive.<init>(JarFileArchive.java:57)
	at org.springframework.xd.module.support.ModuleUtils.createModuleClassLoader(ModuleUtils.java:54)
	at org.springframework.xd.module.support.ModuleUtils.createModuleClassLoader(ModuleUtils.java:47)
	at org.springframework.xd.module.options.DefaultModuleOptionsMetadataResolver.resolveNormalMetadata(DefaultModuleOptionsMetadataResolver.java:186)
	at org.springframework.xd.module.options.DefaultModuleOptionsMetadataResolver.resolve(DefaultModuleOptionsMetadataResolver.java:164)
	at org.springframework.xd.module.options.DelegatingModuleOptionsMetadataResolver.resolve(DelegatingModuleOptionsMetadataResolver.java:44)
	at org.springframework.xd.module.options.EnvironmentAwareModuleOptionsMetadataResolver.resolve(EnvironmentAwareModuleOptionsMetadataResolver.java:127)
	at org.springframework.xd.module.options.DefaultModuleOptionsMetadataResolver.resolveComposedModuleMetadata(DefaultModuleOptionsMetadataResolver.java:175)
	at org.springframework.xd.module.options.DefaultModuleOptionsMetadataResolver.resolve(DefaultModuleOptionsMetadataResolver.java:167)
	at org.springframework.xd.module.options.DelegatingModuleOptionsMetadataResolver.resolve(DelegatingModuleOptionsMetadataResolver.java:44)
	at org.springframework.xd.module.options.EnvironmentAwareModuleOptionsMetadataResolver.resolve(EnvironmentAwareModuleOptionsMetadataResolver.java:127)
	at org.springframework.xd.dirt.stream.XDStreamParser.parse(XDStreamParser.java:174)
```

I think the issue may be related to the following line in ArchiveModuleRegistry:

```
String filename = resource.getFile().getCanonicalFile().getName();
```

**Comment by ilayaperumalg** [comment]
@clemper thanks for reporting. This looks like an issue when using `file:/` based module registry. Created JIRA https://jira.spring.io/browse/XD-3171 to track it. Meanwhile, Can you try using 'hdfs:/' based module registry if that helps in this case? Note that HDFS module registry support was added from 1.2.0.M1 only.

**Comment by clemper** [comment]
Unfortunately, it is not clear to me how to make the change you suggested.  We are using the "xd-yarn push" functionality to deploy the application to the YARN cluster.  How do I enable the hdfs:/ based module registry for both built-in Spring XD modules and our custom modules?  Thanks for your help.

**Comment by trisberg** [comment]
What version of XD are you using? The latest XD is already preconfigured for hdfs as the module registry - https://github.com/spring-projects/spring-xd/blob/master/spring-xd-yarn/site/config/servers.yml#L114

**Comment by ericbottard** [comment]
I'm afraid user is referring to composed modules made out of OOTB modules, which would *not* benefit from the synchronizing feature of the "HDFS backed registry".

The core of the problem seems to be that when running on YARN, each instance is not assured to get its (main) registry backed by the same `file:` path, which is to be expected. The root issue then, is that the deployer is using the path from the serialized module definition (which was likely constructed on the admin), and that does not apply to the container where the module is deployed. If this is really the case, need to make sure that we re-lookup the modules on the running container

**Comment by clemper** [comment]
We encountered this issue on 1.1.2.RELEASE.  We did try upgrading to 1.2.0.RC1 but it did not make any difference.  The composed module is using the internal "transform" processor not a custom module.  We did notice the new "xd.customModule.home" setting but it is not clear how to bulk deploy our custom modules with this setting (as we were able to do with the 1.1.2.RELEASE version by uploading a custom_modules.zip archive).  We could change our custom module deployment process to individually run "module upload" in the shell, but I don't think it would fix the issue with the composed module.

It seems to me that the root of the problem is that the internal module is loaded from the YARN classpath and when serialized into ZK it uses the file path from the admin server (which is then not valid when deserialized on a container running on a different server in the cluster).

**Comment by ilayaperumalg** [comment]
While we investigate on https://jira.spring.io/browse/XD-3171, Can you try using HDFS backed module registry and move the OOTB modules from the default module registry `$XD_HOME/modules` to the HDFS module registry home and see if that works for you?

**Comment by clemper** [comment]
I forked the Spring XD project and modified ZooKeeperComposedModuleDefinitionRegistry to relookup any non-composite module definitions by injecting the "moduleRegistry" bean in the constructor. 

My relookup function looks likes:

```java
private ModuleDefinition relookup(ModuleDefinition definition) {
	if (!definition.isComposed()) {
		return moduleRegistry.findDefinition(definition.getName(), definition.getType());
	}
	List<ModuleDefinition> children = new ArrayList<ModuleDefinition>();
	CompositeModuleDefinition composite = (CompositeModuleDefinition) definition;
	for (ModuleDefinition child : composite.getChildren()) {
		children.add(relookup(child));
	}
	return ModuleDefinitions.composed(
		composite.getName(), composite.getType(), composite.getDslDefinition(), children);
}
```

I call it as follows from findDefintion():

```java
return relookup(this.objectMapper.readValue(new String(data, "UTF-8"),
		ModuleDefinition.class));
```

We are testing this now in our cluster and it does resolve the issue; although, I am not sure that it is a reasonable long term solution.

**Comment by ilayaperumalg** [comment]
Great, please feel free to submit a PR and we can followup. Thanks!

**Comment by ericbottard** [comment]
Created https://github.com/spring-projects/spring-xd/pull/1713

**Comment by clemper** [comment]
I see I wasn't fast enough!  I also created a pull request (https://github.com/spring-projects/spring-xd/pull/1714) which obviously can be ignored, as it is almost identical to yours.  Thanks.

**Comment by ericbottard** [comment]
Well, you definitely deserve all the credit. So I think we'll merge your PR no matter what, with some additional touches here and there (whether you update the PR yourself, or on-merge).

Thanks again for reporting the bug and submitting an initial implementation, your support is appreciated.

**Comment by ilayaperumalg** [comment]
This is addressed in #1714 1714
